File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
  public int maxProduct(int[] nums) {
    int ans = Arrays.stream(nums).max().getAsInt();
    int currMin = 1;
    int currMax = 1;
    int temp;
    for (int num : nums) {
      temp = currMax;
      currMax = Collections.max(Arrays.asList(currMin * num, temp * num, num));
      currMin = Collections.min(Arrays.asList(currMin * num, temp * num, num));
      ans = Collections.max(Arrays.asList(currMax, currMin, ans));
    }
    return ans;
  }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
  def maxProduct(self, nums: List[int]) -> int:
    ans = max(nums)
    currMin, currMax, temp = 1, 1, None
    for n in nums:
      temp = currMax
      currMax = max(currMin * n, temp * n, n)
      currMin = min(currMin * n, temp * n, n)
      ans = max(currMax, currMin, ans)
    return ans